### ISSUE TITLE
Make ivy-purpose-switch-buffer-* detect recentf

### ISSUE THREAD

**Issue by drym3r** [bmag/ivy-purpose#1]
When executing C-x b to change buffer, on the minibuffer appears some closed files (because of ivy-recentf), but when changing to them, it only shows an empty buffer. It should open the closed file.

**Comment by bmag** [comment]
Thanks for the report. I've managed to repro a similar without window-purpose:
- start emacs and make sure all relevant packages are in the load path
- `M-x ivy-mode`
- `M-x recentf-mode`
- open an existing non-empty file, e.g. foo.el
- kill foo.el's buffer (`M-x kill-buffer`)
- `M-x set-variable ivy-use-virtual-buffers t`
- `M-x switch-to-buffer foo.el`: this will open a buffer named foo.el, but it will be empty
- kill foo.el's buffer again
- `M-x ivy-switch-buffer foo.el`: this will open foo.el with the real file's content, as expected

Now, `C-x b` should call `purpose-switch-buffer-overload` (you can check that with `C-h k` or `C-h c`), which defined in window-purpose.el as:
```elisp
(define-purpose-prefix-overload purpose-switch-buffer-overload
  '(purpose-friendly-switch-buffer
    switch-buffer-without-purpose
    purpose-switch-buffer-with-purpose))
```
And `purpose-friendly-switch-buffer` calls either `ido-switch-buffer` or `switch-to-buffer`.

To fix the issue on your side, you can either change the `C-x b` binding or re-define `purpose-switch-buffer-overload`. I'd suggest the first option.
```elisp
;; option 1: unbind window-purpose's `C-x b` binding. This should un-shadow Ivy's binding of `C-x b` to `ivy-switch-buffer`
(define-key purpose-mode-map "C-x b" nil)

;; option 2: redefine purpose-switch-buffer-overload:
(define-purpose-prefix-overload purpose-switch-buffer-overload
  '(ivy-switch-buffer
    switch-buffer-without-purpose
    purpose-switch-buffer-with-purpose))
```

**Comment by drym3r** [comment]
Thanks, the second option worked for me. Seeing the issue you just opened, I see that what you propose in this issue is a hotfix more than a solution, isn't it?

BTW, it also happens with `C-x 4 b`, which executes `purpose-switch-buffer-other-window-overload`.

**Comment by bmag** [comment]
> what you propose in this issue is a hotfix more than a solution

That's correct, and I need to think how to solve it properly. The easiest way is probably to get rid of the whole prefix-overloading stuff and not replace any of the standard key bindings. It will take me some time to sort this all out. (unless Ivy will stop adding virtual buffers to `switch-to-buffer`'s candidates, but I suspect there are some difficult-to-overcome technical subtleties in play there)

>  it also happens with `C-x 4 b`

And probably `C-x 5 b` too. You can find the definitions of `purpose-switch-buffer-other-window-overload` and `purpose-switch-buffer-other-frame-overload` in `window-purpose.el`, and re-define them in a similar manner to `purpose-switch-buffer-overload`. However, `ivy-switch-buffer` can also open a buffer in another window (there's an action for it), so you probably don't need `C-x 4 b`. As for `C-x 5 b`, I'm not aware if Ivy has a feature for opening buffers in another frame (but it can be extended).

**Comment by drym3r** [comment]
It only seems to be `ivy-switch-buffer` and `ivy-switch-buffer-other-window`. Anyway, thanks for the effort and for this package, I 'll wait until you sort it. Sorry I can't help much more.

**Comment by Kungsgeten** [comment]
This version of the second solution worked better for me:

    (defun ivy-switch-buffer-and-select-window ()
      "Use `ivy-switch-buffer' and then select the buffer's window."
      (interactive)
      (select-window (get-buffer-window (ivy-switch-buffer))))
    (define-purpose-prefix-overload purpose-switch-buffer-overload
      '(ivy-switch-buffer-and-select-window
        ivy-purpose-switch-buffer-without-purpose
        ivy-purpose-switch-buffer-with-purpose))

Now it will switch to the window containing the chosen buffer, which it wouldn't (at least not for me) when using `ivy-switch-buffer`.

**Comment by cdgreenidge** [comment]
I chose to use option 1, but needed to change it to

```
(define-key purpose-mode-map (kbd "C-x b") nil)
```

to get it to work.
